lib/ansible/modules/cloud/amazon/ec2_vol.py

```diff
@@ -373,117 +386,196 @@ def attach_volume(module, ec2, volume, instance):
+            VolumeId=volume_dict['volume_id'],
+        )
+        waiter = ec2_conn.get_waiter('volume_in_use')
+        waiter.wait(
+            VolumeIds=[attach_response['VolumeId']],
+        )
+        changed = True
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg='Error while attaching EBS volume')
 
-        modify_dot_attribute(module, ec2, instance, device_name)
+    modify_dot_attribute(module, ec2_conn, instance_dict, device_name)
 
+    volume = get_volume(module, ec2_conn, vol_id=volume_dict['volume_id'])
     return volume, changed
 
 
-def modify_dot_attribute(module, ec2, instance, device_name):
+def modify_dot_attribute(module, ec2_conn, instance_dict, device_name):
     """ Modify delete_on_termination attribute """
 
     delete_on_termination = module.params.get('delete_on_termination')
     changed = False
 
-    try:
-        instance.update()
-        dot = instance.block_device_mapping[device_name].delete_on_termination
-    except boto.exception.BotoServerError as e:
-        module.fail_json(msg="%s: %s" % (e.error_code, e.error_message))
+    instance_dict = get_instance(module, ec2_conn=ec2_conn, instance_id=instance_dict['instance_id'])
+    mapped_block_device = get_mapped_block_device(instance_dict=instance_dict, device_name=device_name)
 
-    if delete_on_termination != dot:
+    if delete_on_termination != mapped_block_device['ebs']['delete_on_termination']:
         try:
-            bdt = BlockDeviceType(delete_on_termination=delete_on_termination)
-            bdm = BlockDeviceMapping()
-            bdm[device_name] = bdt
-
-            ec2.modify_instance_attribute(instance_id=instance.id, attribute='blockDeviceMapping', value=bdm)
-
-            while instance.block_device_mapping[device_name].delete_on_termination != delete_on_termination:
-                time.sleep(3)
-                instance.update()
+            ec2_conn.modify_instance_attribute(
+                InstanceId=instance_dict['instance_id'],
+                BlockDeviceMappings={
+                    "DeviceName": device_name,
+                    "Ebs": {
+                        "DeleteOnTermination": delete_on_termination
+                    }
+                }
+            )
+
+            # while instance.block_device_mapping[device_name].delete_on_termination != delete_on_termination:
+            #     time.sleep(3)
+            #     instance.update()
             changed = True
-        except boto.exception.BotoServerError as e:
-            module.fail_json(msg="%s: %s" % (e.error_code, e.error_message))
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e,
+                                 msg='Error while modifiying Block Device Mapping of instances',
+                                 instance_id=instance_dict['instance_id'])
```

Comment:
Maybe just add the ID of the instance to the msg. `'Error while modifiying Block Device Mapping of instance {0}'.format(instance_dict['instance_id'])`